stdlib/lazy.mli

```diff
@@ -67,36 +66,76 @@ external force : 'a t -> 'a = "%lazy_force"
 
-    If the computation of [x] raises an exception, it is unspecified
-    whether [force_val x] raises the same exception or {!Undefined}.
-    @raise Undefined if the forcing of [x] tries to force [x] itself
-    recursively.
+(** {1 Iterators} **)
+
+val map : ('a -> 'b) -> 'a t -> 'b t
+(** [map f x] returns a suspension that, when forced,
+    forces [x] and applies [f] to its value.
+
+    It is equivalent to [lazy (f (Lazy.force x))].
+
+    @since 4.13.0
 *)
 
+(** {1 Reasoning on already-forced suspensions} *)
+
+val is_val : 'a t -> bool
+(** [is_val x] returns [true] if [x] has already been forced and
+    did not raise an exception.
+    @since 4.00.0 *)
+
+val from_val : 'a -> 'a t
+(** [from_val v] evaluates [v] first (as any function would) and returns
+    an already-forced suspension of its result.
+    It is the same as [let x = v in lazy x], but uses dynamic tests
+    to optimize suspension creation in some cases.
+    @since 4.00.0 *)
+
+val map_val : ('a -> 'b) -> 'a t -> 'b t
+(** [map_val f x] applies [f] directly if [x] is already forced,
+   otherwise it behaves as [map f x].
+
+   When [x] is already forced, this behavior saves the construction of
+   a suspension, but on the other hand it performs more work eagerly
+   that may not be useful if you never force the function result.
+
+   If [f] raises an exception, it will be raised immediately when
+   [is_val x], or raised only when forcing the thunk otherwise.
+
+   If [map_val f x] returns a non-exception result, then
+   [is_val (map_val f x)] is equal to [is_val x].
+
+    @since 4.13.0 *)
+
+
+(** {1 Advanced}
+
+   The following definitions are for advanced uses only; they require
+   familiary with the lazy compilation scheme to be used appropriately. *)
+
 val from_fun : (unit -> 'a) -> 'a t
 (** [from_fun f] is the same as [lazy (f ())] but slightly more efficient.
 
-    [from_fun] should only be used if the function [f] is already defined.
+    It should only be used if the function [f] is already defined.
     In particular it is always less efficient to write
     [from_fun (fun () -> expr)] than [lazy expr].
 
     @since 4.00.0 *)
 
-val from_val : 'a -> 'a t
-(** [from_val v] returns an already-forced suspension of [v].
-    This is for special purposes only and should not be confused with
-    [lazy (v)].
-    @since 4.00.0 *)
+val force_val : 'a t -> 'a
+(** [force_val x] forces the suspension [x] and returns its
+    result. If [x] has already been forced, [force_val x]
+    returns the same value again without recomputing it.
 
-val is_val : 'a t -> bool
-(** [is_val x] returns [true] if [x] has already been forced and
-    did not raise an exception.
-    @since 4.00.0 *)
+    If the computation of [x] raises an exception, it is unspecified
+    whether [force_val x] raises the same exception or {!Undefined}.
+    @raise Undefined if the forcing of [x] tries to force [x] itself
+    recursively.
```

Comment:
This docstring notes some unspecified behaviour, and then specifies it in the next sentence.  It seems fairly clear under what circumstances force_val will raise `Undefined`.

Reply:
I think that the two sentences are talking about different things: (1) what happens if the suspended computation raises an exception (say `lazy (raise Exit)`), and (2) what happens if the thunk tries to force itself. In the second case it (claims that it) raises Undefined, in the first case it may raise Undefined or re-raise the exception.

In any case, this is unrelated to the change in this PR, so I would propose to deal with any potential change proposals here in a separate issue/PR.

Reply:
Ah yes, that's clearer. I guess I don't understand under what circumstances it's ever safe to raise an exception from within something forced by this function then, as it always has the opportunity to gobble the exception and raise a Undefined instead.  Give that, how can an exception ever safely be thrown?

Reply:
(There's the case of a non-recursive function for which you can inspect all possible exceptions, but then we already know that it can never raised `Undefined` from the next sentence. But the specification of force_val retains the possibility of raising an Undefined even in that case).

Reply:
I think this `force_val` function is intended to be used for thunks that we know are already forced (for example by checking `is_val` first), and those thunks will not raise an exception. It's in the Advanced section, so probably not intended for wide usage.

Reply:
I can't tell about the original intent, but `force_val` is at most no better than `force` for already forced chunks (it's usually less efficient because `force` is a compiler primitive whereas `force_val` isn't).
It's only marginally better for not-yet-forced thunks because it skips one try-with handler, and unless AFL instrumentation is on (that prevents some optimisations from triggering) it probably still ends up slower anyway.